File: DATA_BINDING/app/src/main/java/junseong/android/com/MainActivity.java
package junseong.android.com;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import java.util.List;

import junseong.android.com.databinding.ActivityMainBinding;
import junseong.android.com.databinding.ItemMainBinding;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    private static final String QUERY = "movies";
    private static final String API_KEY = "ebbc3f52dc294e72aa888b8fe88aa962";

    ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding= DataBindingUtil.setContentView(this,R.layout.activity_main);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));

        RetrofitService networkService = RetrofitFactory.create();

        networkService.getList(QUERY, API_KEY, 1, 10)
                .enqueue(new Callback<PageListModel>() {
                    @Override
                    public void onResponse(Call<PageListModel> call, Response<PageListModel> response) {
                        if (response.isSuccessful()) {
                            MyAdapter adapter=new MyAdapter(response.body().articles);
                            binding.recyclerView.setAdapter(adapter);
                        }
                    }

                    @Override
                    public void onFailure(Call<PageListModel> call, Throwable t) {

                    }
                });

    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
       ItemMainBinding binding;
       public ItemViewHolder(ItemMainBinding binding){
           super(binding.getRoot());
           this.binding=binding;
       }
    }

    class MyAdapter extends RecyclerView.Adapter<ItemViewHolder> {
        List<ItemModel> articles;

        public MyAdapter(List<ItemModel> articles) {
            this.articles = articles;
        }

        @Override
        public int getItemCount() {
            return articles.size();
        }

        @NonNull
        @Override
        public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            ItemMainBinding binding=ItemMainBinding.inflate(LayoutInflater.from(parent.getContext()),parent,false);
            return new ItemViewHolder(binding);
        }

        @Override
        public void onBindViewHolder(@NonNull ItemViewHolder holder, int position) {
            ItemModel model=articles.get(position);
            holder.binding.setItem(model);
        }
    }
    @BindingAdapter("bind:publishedAt")
    public static void publishedAt(TextView view,String date){
        view.setText(AppUtils.getDate(date)+"at"+AppUtils.getTime(date));
    }
    @BindingAdapter("bind:urlToImage")
    public static void urlToImage(ImageView view,String url){
        Glide.with(MyApplication.getAppContext()).load(url).override(250,250).into(view);
    }
}

File: customAdpater(Kotlin)/app/src/main/java/junseong/kotlin/customadpater/MainActivity.kt
package junseong.kotlin.customadpater

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.row.view.*

class MainActivity : AppCompatActivity() {
    val data1=Array(10){""}
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        for ( i in 0 until 10)
            data1[i]="데이터${i+1}"

        list.adapter=adapter1
    }
    val adapter1=object  :BaseAdapter()  {
        override fun getView(position: Int, p1: View?, p2: ViewGroup?): View {
            var rowView=p1
            if(rowView==null){
                rowView=layoutInflater.inflate(R.layout.row,null)
            }
            /*
            //항목 뷰 내부에 배치되어 있는 뷰들의 주소를 가져옴 첫번째 방법
            val tmptext=rowView?.findViewById<TextView>(R.id.rowtext)
            val btn1=rowView?.findViewById<Button>(R.id.button1)
            val btn2=rowView?.findViewById<Button>(R.id.button2)
            tmptext?.text=data1[position]

            btn1?.tag=position+1
            btn2?.tag=position+1


            btn1?.setOnClickListener{
                text1.text="첫번째 버튼누름 ${it.tag}"
            }
            btn2?.setOnClickListener{
                text1.text="두번째버튼누름 ${it.tag}"
            }

             */

            /* 두번째 방법 */
            rowView?.run{
                rowtext.text=data1[position]
                button1.tag=position+1
                button2.tag=position+1
                button1.setOnClickListener {
                 text1.text="첫번째 버튼을 누름 : ${it.tag}"
                }
                button2.setOnClickListener {
                    text1.text="두번째 버튼을 누름 :${it.tag}"
                }
            }
            return rowView!!
        }

        override fun getItem(p0: Int): Any? {
            return null
        }

        override fun getItemId(p0: Int): Long {
           return 0
        }

        override fun getCount(): Int {
            return data1.size
        }
    }
}
